### PR TITLE
Implementation of C51, QR-DQN, IQN

### PR DESCRIPTION
**Pull request by liyuanpeng01**
Pull request with clean commit history.

### CODE REVIEW

**Comment by yu239** [comment]
Please refer to PR #26 for my comments.

File: flare/model_zoo/tests/test_distributional_rl_models.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        self.assertEqual((batch_size, num_actions), q_value.size())
+        for i in xrange(batch_size):
+            for j in xrange(num_actions):
+                self.assertAlmostEqual(
+                    1.,
+                    sum(q_value_distribution[i][j].data.tolist()),
+                    places=6)
+
+
+class TestSimpleSimpleModelQRDQN(unittest.TestCase):
```
Comment (yu239) [inline flare/model_zoo/tests/test_distributional_rl_models.py]: TestSimpleModelQRDQN

File: flare/examples/img_q_example.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
+import torch.nn as nn
```
Comment (yu239) [inline flare/examples/img_q_example.py]: Please make sure that each example can produce a reasonable training curve.

File: flare/examples/distributional_rl/c51_example.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
+import torch.nn as nn
```
Comment (yu239) [inline flare/examples/distributional_rl/c51_example.py]: Maybe consider integrating all the similar examples into one single file and let the user specify the algorithm/model. Currently too much repeated code.

**Comment by liyuanpeng01** [comment]
Learning curve for C51.
Horizontal axis is number of total games (x100).
Vertical axis is game reward.
The curve is similar to the left part of AC curve in README.md file.
c51

Also for Q Learning.
q

QR-DQN and IQN results will be posted later.

File: flare/examples/distributional_rl/c51_example.py (hunk 3 above)
Comment (liyuanpeng01) [reply flare/examples/distributional_rl/c51_example.py]: Example files are merged.

File: flare/examples/img_q_example.py (hunk 2 above)
Comment (liyuanpeng01) [reply flare/examples/img_q_example.py]: The produced curves seem comparable with the left part of AC curve.
